x/concentrated-liquidity/fees_test.go

```diff
@@ -1015,3 +1020,152 @@ func (s *KeeperTestSuite) TestPreparePositionAccumulator() {
 		})
 	}
 }
+
+func (s *KeeperTestSuite) TestFunctionalFees() {
+	// Init suite.
+	s.Setup()
+
+	// Default setup only creates 3 accounts, but we need 5 for this test.
+	s.TestAccs = apptesting.CreateRandomAccounts(5)
+
+	// Create a default CL pool, but with a 10 percent swap fee.
+	clPool := s.PrepareCustomConcentratedPool(s.TestAccs[0], ETH, USDC, DefaultTickSpacing, DefaultExponentAtPriceOne, sdk.MustNewDecFromStr("0.1"))
+
+	// Setup full range position across all four accounts
+	for i := 0; i < 4; i++ {
+		s.SetupFullRangePositionAcc(clPool.GetId(), s.TestAccs[i])
+	}
+
+	// Setup narrow range position across three of four accounts
+	for i := 0; i < 3; i++ {
+		s.SetupDefaultPositionAcc(clPool.GetId(), s.TestAccs[i])
+	}
+
+	// Setup consecutive range position (in relation to narrow range position) across two of four accounts
+	for i := 0; i < 2; i++ {
+		s.SetupConsecutiveRangePositionAcc(clPool.GetId(), s.TestAccs[i])
+	}
+
+	// Setup overlapping range position (in relation to narrow range position) on one of four accounts
+	s.SetupOverlappingRangePositionAcc(clPool.GetId(), s.TestAccs[0])
+
+	// Swap multiple times USDC for ETH, therefore increasing the spot price
+	ticksActivatedAfterEachSwap, totalFeesExpected := s.swapAndTrackXTimesInARow(clPool.GetId(), DefaultCoin1, ETH, cltypes.MaxSpotPrice, 7)
+	s.CollectAndAssertFees(s.Ctx, clPool.GetId(), totalFeesExpected, [][]sdk.Int{ticksActivatedAfterEachSwap}, onlyUSDC)
+
+	// Swap multiple times ETH for USDC, therefore decreasing the spot price
+	ticksActivatedAfterEachSwap, totalFeesExpected = s.swapAndTrackXTimesInARow(clPool.GetId(), DefaultCoin0, USDC, cltypes.MinSpotPrice, 7)
+	s.CollectAndAssertFees(s.Ctx, clPool.GetId(), totalFeesExpected, [][]sdk.Int{ticksActivatedAfterEachSwap}, onlyETH)
+
+	// Do the same swaps as before, however this time we collect fees after both swap directions are complete.
+	ticksActivatedAfterEachSwapUp, totalFeesExpectedUp := s.swapAndTrackXTimesInARow(clPool.GetId(), DefaultCoin1, ETH, cltypes.MaxSpotPrice, 7)
+	ticksActivatedAfterEachSwapDown, totalFeesExpectedDown := s.swapAndTrackXTimesInARow(clPool.GetId(), DefaultCoin0, USDC, cltypes.MinSpotPrice, 7)
+	totalFeesExpected = totalFeesExpectedUp.Add(totalFeesExpectedDown...)
+
+	// We expect all positions to have both denoms in their fee accumulators except USDC for the overlapping range position since
+	// it was not activated during the USDC -> ETH swap direction but was activated during the ETH -> USDC swap direction.
+	ticksActivatedAfterEachSwapTest := [][]sdk.Int{ticksActivatedAfterEachSwapUp, ticksActivatedAfterEachSwapDown}
+	denomsExpected := [][]string{{USDC, ETH}, {USDC, ETH}, {USDC, ETH}, {"", ETH}}
```

Comment:
What does an empty string mean?

Reply:
Can you elaborate with a comment please?

Reply:
No longer use an empty string directly so the const should explain now via its name https://github.com/osmosis-labs/osmosis/pull/4282/commits/c1a61b883c5abdc399a11108c5fc05d65a6083a8